controllers/apim/virtualservice_controller.go

```diff
@@ -0,0 +1,216 @@
+		},
+	}
+
+	providerName, present := vs.GetAnnotations()[KuadrantAuthProviderAnnotation]
+	if !present {
+		return fmt.Errorf("Kuadrant auth-provider annotation not found")
+	}
+
+	// fill out the rules
+	authToRules := []*securityv1beta1.Rule_To{}
+	for _, httpRoute := range vs.Spec.Http {
+		for idx, matchRequest := range httpRoute.Match {
+			toRule := &securityv1beta1.Rule_To{
+				Operation: &securityv1beta1.Operation{},
+			}
+
+			toRule.Operation.Hosts = vs.Spec.Hosts
+			if normalizedURI := normalizeStringMatch(matchRequest.Uri); normalizedURI != "" {
+				toRule.Operation.Paths = append(toRule.Operation.Paths, normalizedURI)
+			}
+
+			if normalizedMethod := normalizeStringMatch(matchRequest.Method); normalizedMethod != "" {
+				// Looks like it's case-sensitive:
+				// https://istio.io/latest/docs/reference/config/security/normalization/#1-method-not-in-upper-case
+				method := strings.ToUpper(normalizedMethod)
+				toRule.Operation.Methods = append(toRule.Operation.Methods, method)
+			}
+
+			// If there is only regex stringmatches then we'll have bunch of repeated To rules with
+			// only same host filled into each. Following make sure only one field like that is present.
+			operation := toRule.Operation
+			if len(operation.Paths) == 0 && len(operation.Methods) == 0 && idx > 0 {
+				continue
+			}
+			authToRules = append(authToRules, toRule)
+		}
+	}
+
+	authPolicy.Spec = securityv1beta1.AuthorizationPolicy{
+		Rules: []*securityv1beta1.Rule{{
+			To: authToRules,
+		}},
+		Action: securityv1beta1.AuthorizationPolicy_CUSTOM,
+		ActionDetail: &securityv1beta1.AuthorizationPolicy_Provider{
+			Provider: &securityv1beta1.AuthorizationPolicy_ExtensionProvider{
+				Name: providerName,
+			},
+		},
+	}
+
+	if err := controllerutil.SetOwnerReference(vs, &authPolicy, r.Client().Scheme()); err != nil {
+		logger.Error(err, "failed to add owner ref to AuthorizationPolicy resource")
+		return err
+	}
+	err := r.ReconcileResource(ctx, &istiosecurityv1beta1.AuthorizationPolicy{}, &authPolicy, alwaysUpdateAuthPolicy)
+	if err != nil && !apierrors.IsAlreadyExists(err) {
+		logger.Error(err, "ReconcileResource failed to create/update AuthorizationPolicy resource")
+		return err
+	}
+
+	logger.Info("successfully created/updated AuthorizationPolicy resources")
+	return nil
+}
+
+func alwaysUpdateAuthPolicy(existingObj, desiredObj client.Object) (bool, error) {
+	existing, ok := existingObj.(*istiosecurityv1beta1.AuthorizationPolicy)
+	if !ok {
+		return false, fmt.Errorf("%T is not a *istiosecurityv1beta1.AuthorizationPolicy", existingObj)
+	}
+	desired, ok := desiredObj.(*istiosecurityv1beta1.AuthorizationPolicy)
+	if !ok {
+		return false, fmt.Errorf("%T is not a *istiosecurityv1beta1.AuthorizationPolicy", desiredObj)
+	}
+
+	existing.Spec = desired.Spec
+	return true, nil
+}
+
+func normalizeStringMatch(sm *networkingv1alpha3.StringMatch) string {
+	if prefix := sm.GetPrefix(); prefix != "" {
+		return prefix + "*"
```

Comment:
why do we need to add this? Is it due to a difference between VS and AuthorizationPolicy? Does prefix always end with ```/``` or is it valid to have ```/cars*```?

Reply:
Yes, due to differences between them. From the docs `Prefix match: “abc*” will match on value “abc” and “abcd”.`, it looks like it will be valid.